### DISCUSSION TITLE
Redis and Redis Sentinel - How to add username, password and database index to DSN

### DISCUSSION THREAD

**Discussion by ToshY** [symfony/symfony#47361] [Q&A] [answered]
I'm having a hard time understanding the several different possible DSN formats for Redis and Redis Sentinel. 

---

First let me start off with my normal usage with a single Redis instance (using docker).

```dotenv
redis://${REDIS_USER}:${REDIS_PASSWORD}@redis:6379/3
```
This allows me to specify the user `${REDIS_USER}`, password `${REDIS_PASSWORD}`, host `redis`, port `6379` and database index `3`.

This seems the simplest of DSN format for a single Redis instance and works like a charm :+1:

---

Now comes a (subjectively) more "menacing" format. According to the [docs](https://symfony.com/doc/current/components/cache/adapters/redis_adapter.html#configure-the-connection) for multiple hosts:

```dotenv
redis:?host[localhost]&host[localhost:6379]&host[/var/run/redis.sock:]&auth=my-password&redis_cluster=1
```

Disregarding the query parameter `redis_cluster=1`, in theory it means I should be able to apply this format for a single Redis instance as well (?)
If so, how would I specify a **username**, **password** and **dbindex** in that DSN?

I've tried several combinations of DSN's, of which all of the following failed:

```dotenv
redis:?host[redis:6379]]&auth=${REDIS_PASSWORD} - "NOAUTH Authentication required"
redis:?host[redis:6379/3]]&auth=${REDIS_PASSWORD} - "NOAUTH Authentication required" 
redis:?host[${REDIS_USERNAME}:${REDIS_PASSWORD}@redis:6379/3] - Failed to parse IPv6 address "[redis:6379/3]]:6379"
redis:?host[${REDIS_USERNAME}:${REDIS_PASSWORD}@redis:6379]/3 - Failed to parse IPv6 address "[redis:6379]]:6379"
redis:?host[${REDIS_USERNAME}:${REDIS_PASSWORD}@redis:6379]&dbindex=3 - Failed to parse IPv6 address "[redis:6379]&dbindex=3]:6379"
redis:?host[redis:6379]&user=${REDIS_USER}&password=${REDIS_PASSWORD}&dbindex=3 - NOAUTH Authentication required.
```

Think it's clear that at some point I'm not sure what I'm even doing anymore, and the documentation isn't helping here,  as it's pointing to the [RedisTrait](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Cache/Traits/RedisTrait.php), which is something I don't know if I want to try and understands what's going on 😨.

---

For Redis Sentinel the following example is shown in the same doc paragraph:

```dotenv
redis:?host[redis1:26379]&host[redis2:26379]&host[redis3:26379]&redis_sentinel=mymaster
```

Having setup 1 master (container: `redis-master`; monitor name `redismaster`), 1 slave (container: `redis-slave`) and 3 sentinels (containers: `redis-sentinel-<1-3>`; ports: `26380-26382:26379` (outside:inside)), I imagine it looking like this:

```dotenv
redis:?host[redis-sentinel-1:26379]&host[redis-sentinel-2:26379]&host[redis-sentinel-3:26379]&redis_sentinel=redismaster
```

Same question: How do I specify a **username**, **password** and **dbindex** in that DSN?

---

**Edit**

@Nyholm As you seem to have updated the documentation in [this commit](https://github.com/symfony/symfony-docs/commit/39a24f75cc0584ae01a2d4a266bd1873c122eaf0), could you care to elaborate on what the correct syntax would be for adding username, password and database index for both a single Redis instance as well as well as Sentinel?

@nicolas-grekas As you've worked on the [RedisTrait](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Cache/Traits/RedisTrait.php), could you possibly give more insight on the DSN formats?

**Comment by MatTheCat** [comment]
So I tried to understand the  `RedisTrait` :grin: 

```
redis:?host[redis:6379]&user=${REDIS_USER}&password=${REDIS_PASSWORD}&dbindex=3
```

For a single host the dbindex is the DSN path and auth is done as for HTTP so you would have

```
redis://${REDIS_USER}:${REDIS_PASSWORD}@redis/3
```

(Note that you must have Symfony ≥ 6.1 to specify a user.)

When using multiple hosts syntax I think you have to put each of their parameters in their query string, so I guess you could write

```
redis:?host[redis?auth=${REDIS_USER}:${REDIS_PASSWORD}&dbindex=3]
```

(I removed the port because it is the default one.)

**Reply by ToshY** [reply]
hey @MatTheCat 👋 

Thanks for your response. I just tried this and it gave me `Invalid Redis DSN: "redis:?host[redis?auth=default:password&dbindex=3]`. (`default:password` are my testing credentials for user/password in this context).

p.s.

I'm on Symfony 6.1.3

**Reply by MatTheCat** [reply]
Damn 🤔

What about

```
redis:?host[redis]=auth%3Ddefault:password%26dbindex%3D3
```
 ?

**Reply by ToshY** [reply]
Okay so I've tried it, and it gets even stranger now... 🤔 

But before that, I'll give you a bit of background information of what I'm using Redis for. 

I'm using Redis to store pre-generated [blurhashes](https://blurha.sh/) of images. These blurred images are just sitting in Redis (dbindex `3`). Normally, if the database is empty, and I would load the homepage, it will take +/- 20-25 seconds to get the +/- 130 images from external storage, generate blurhashes, put them in Redis, and then eventually show the rendered homepage to the client. Initially that's slow (that's why I pre-generate them normally), but now that they are saved in Redis, in the next request the homepage will render within 500-600 ms.

So when I now look up the keys in Redis, I see 130 keys, denoting the 130 blurhashes it just generated. This is all with the "normal" DSN format.

---

After changing the DSN to your last suggestion, it took a "while" to load. To be more specific, it eventually reached the maximum execution time of 30 seconds, but managed to load the page partially and show the debug message within an image's container div. 

```
# First try
Fatal error: Maximum execution time of 30 seconds exceeded in 
/app/vendor/intervention/image/src/Intervention/Image/Gd/Decoder.php:128

#Second try
Fatal error: Maximum execution time of 30 seconds exceeded in 
/app/vendor/kornrunner/blurhash/src/Blurhash.php:120
```

```
# Complete trace (which doesn't really tell me anything)

Error:
Typed property Symfony\Component\Cache\Adapter\TraceableAdapterEvent::$end must not be accessed before initialization

  at /app/vendor/symfony/cache/DataCollector/CacheDataCollector.php:115
  at Symfony\Component\Cache\DataCollector\CacheDataCollector-&gt;calculateStatistics()
     (/app/vendor/symfony/cache/DataCollector/CacheDataCollector.php:50)
  at Symfony\Component\Cache\DataCollector\CacheDataCollector-&gt;collect(object(Request), object(Response), object(FatalError))
     (/app/vendor/symfony/http-kernel/Profiler/Profiler.php:156)
  at Symfony\Component\HttpKernel\Profiler\Profiler-&gt;collect(object(Request), object(Response), object(FatalError))
     (/app/vendor/symfony/http-kernel/EventListener/ProfilerListener.php:108)
  at Symfony\Component\HttpKernel\EventListener\ProfilerListener-&gt;onKernelResponse(object(ResponseEvent), 'kernel.response', object(TraceableEventDispatcher))
     (/app/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:115)
  at Symfony\Component\EventDispatcher\Debug\WrappedListener-&gt;__invoke(object(ResponseEvent), 'kernel.response', object(TraceableEventDispatcher))
     (/app/vendor/symfony/event-dispatcher/EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher-&gt;callListeners(array(object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener)), 'kernel.response', object(ResponseEvent))
     (/app/vendor/symfony/event-dispatcher/EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher-&gt;dispatch(object(ResponseEvent), 'kernel.response')
     (/app/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:153)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher-&gt;dispatch(object(ResponseEvent), 'kernel.response')
     (/app/vendor/symfony/http-kernel/HttpKernel.php:185)
  at Symfony\Component\HttpKernel\HttpKernel-&gt;filterResponse(object(Response), object(Request), 2)
     (/app/vendor/symfony/http-kernel/HttpKernel.php:173)
  at Symfony\Component\HttpKernel\HttpKernel-&gt;handleRaw(object(Request), 2)
     (/app/vendor/symfony/http-kernel/HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel-&gt;handle(object(Request), 2, false)
     (/app/vendor/symfony/http-kernel/EventListener/ErrorListener.php:85)
  at Symfony\Component\HttpKernel\EventListener\ErrorListener-&gt;onKernelException(object(ExceptionEvent), 'kernel.exception', object(TraceableEventDispatcher))
     (/app/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:115)
  at Symfony\Component\EventDispatcher\Debug\WrappedListener-&gt;__invoke(object(ExceptionEvent), 'kernel.exception', object(TraceableEventDispatcher))
     (/app/vendor/symfony/event-dispatcher/EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher-&gt;callListeners(array(object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener)), 'kernel.exception', object(ExceptionEvent))
     (/app/vendor/symfony/event-dispatcher/EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher-&gt;dispatch(object(ExceptionEvent), 'kernel.exception')
     (/app/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:153)
  at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher-&gt;dispatch(object(ExceptionEvent), 'kernel.exception')
     (/app/vendor/symfony/http-kernel/HttpKernel.php:213)
  at Symfony\Component\HttpKernel\HttpKernel-&gt;handleThrowable(object(FatalError), object(Request), 1)
     (/app/vendor/symfony/http-kernel/HttpKernel.php:106)
  at Symfony\Component\HttpKernel\HttpKernel-&gt;terminateWithException(object(FatalError), object(Request))
     (/app/vendor/symfony/http-kernel/EventListener/DebugHandlersListener.php:125)
  at Symfony\Component\HttpKernel\EventListener\DebugHandlersListener::Symfony\Component\HttpKernel\EventListener\{closure}(object(FatalError))
     (/app/vendor/symfony/error-handler/ErrorHandler.php:540)
  at Symfony\Component\ErrorHandler\ErrorHandler-&gt;handleException(object(FatalError))
     (/app/vendor/symfony/error-handler/ErrorHandler.php:630)
  at Symfony\Component\ErrorHandler\ErrorHandler::handleFatalError()                
```

Okay, so my initial thoughts/findings were the following with your suggested DSN format:

* The files were they time-out occurs, while being used to generate blurhashes, never had issues with the old DSN.  So while it times out in those files, the cause is most likely somewhere else: most likely the new DSN format.
* Looking in the Redis database (using Redis Insight v2), I looked at database index `3`, and surprisingly it was still empty. Nothing was written to it.
* Refreshing the page the second time also gave the same stacktrace regarding the typed property issue and was equally slow as the first time. (this should've been the first indication that images were not being saved in Redis).

---

While it _seems_ that the DSN is regarded valid (as I don't see get an exception thrown related to the validity), for some reason it doesn't write anything to Redis 🤔 So that makes me doubt if it's actually valid and adds even more confusion to the whole problem.

**Reply by MatTheCat** [reply]
Ah yes `auth` is supposed to be an array. So I guess `default:password` was interpreted as a password… let’s try

```
redis:?host[redis]=auth%5B%5D%3Ddefault%26auth%5B%5D%3Dpassword%26dbindex%3D3
```

**Reply by ToshY** [reply]
Doesn't work, gives `NOAUTH` again. Also, I'm not sure why you suggest to an URL encoded string, as none of the examples use that? And the decoded DSN doesn't seem valid: `redis:?host[redis]=auth[]=default&auth[]=password&dbindex=3`. (emphasis on `[redis]=auth[]`)

Also the following examples also do not work:

```
redis:?host[redis]&auth[]=default&auth[]=password&dbindex=3
redis:?host[redis]&auth['user']=default&auth['password']=password&dbindex=3
redis:?host[redis]&auth['username']=default&auth['password']=password&dbindex=3
```

At this point I'm not sure if it's even possible... 😞

**Reply by ToshY** [reply]
@MatTheCat I've edited the question and tagged Nyholm, which seems to have added the documentation for the other DSN format. In the off chance he sees this, maybe he could give some clarification on the proper syntax 🤞

**Comment by ToshY** [comment, marked as answer]
I finally found the correct syntax in the [symfony docs issue 16310](https://github.com/symfony/symfony-docs/issues/16310). 

For a single Redis instance: 

```
redis:${REDIS_USER}:${REDIS_PASSWORD}@?host[redis:6379]&dbindex=3
```

For sentinel:

```
redis:${REDIS_USER}:${REDIS_PASSWORD}@?host[redis-sentinel-redis-sentinel-1:26379]&host[redis-sentinel-redis-sentinel-3:26379]&host[redis-sentinel-redis-sentinel-3:26379]&redis_sentinel=redismaster&dbindex=3
```

<sup>Note: the `redis-sentinal-redis-sentinel-1` syntax is due to running the docker compose service `redis-sentinel` in another folder called `redis-sentinel`, thus making the container names `redis-sentinel-redis-sentinel-<1,2,3>`</sup>

For Sentinel, I was also required to install [Predis](https://github.com/predis/predis) and use that client instead of `Redis`, even though the docs note that having the Redis extension 5.2+ should have sufficed (I had 5.3.7).

@MatTheCat Thanks for the help 👍 I will make a PR for the docs to update the DSN section 😄
